projects/feature-flag.md

```diff
@@ -0,0 +1,141 @@
+
+This project also includes Dan Dyla acting as one of the project leads.
+Dan is a member of the GC, maintainer of OTel JS, and has experience developing semantic conventions.
+Dan will provide the necessary expertise to ensure the proposed semantic conventions meet the standards and expectations of the OpenTelemetry community and semantic conventions SIG.
+
+The project is sponsored by Dan Dyla and Alexander Wert, an OpenTelemetry semantic conventions maintainer.
+
+### Required staffing
+
+**Project Leads:** Daniel Dyla and Michael Beemer (OpenFeature co-founder/GC member)
+
+**Project Sponsors**: Daniel Dyla and Alexander Wert
+
+**Engineers willing to write and review prototypes:**
+
+TODO: list prototype languages and engineers
+
+The following is a list of engineers committed to working on the project:
+
+- JS Prototype - Daniel Dyla (OTel GC, OTel JS, Dynatrace)
+- Review Prototypes - Ryan Lamb (OpenFeature TC, LaunchDarkly)
+- TODO: more engineers
+
+## Meeting Times
+
+This project expects to meet every 2 weeks until this charter is fulfilled.
+The exact time and date of the meeting will be determined after the project is approved.
+
+## Timeline
+
+We expect the project to take about 1 calendar quarter. The OpenFeature project is currently targeting Kubecon NA 2024 to announce stable OpenTelemetry support.
+
+- 2-4 weeks - First experimental semantic convention built on pre-existing feature flag semantic conventions
+- 2-4 weeks - Prototype implementation and review
+- 2-4 weeks - Incorporate feedback from prototype implementers
+- 2-4 weeks - Final review period including the greater spec and maintainer community
+
+## Labels
+
+This project will use the label [`area:feature-flag`](https://github.com/open-telemetry/semantic-conventions/labels/area%3Afeature-flag) to track issues and pull requests in the semantic conventions repository.
+
+## Project Board
+
+Once the project is approved a project board will be created and linked here.
+The project lead and all other relevant project members should have edit access to the board.
+The board will be pre-populated with at least the following issues to track deliverables and the timeline.
+
+- **signal type for feature flag impressions** - agree on which signal type or types should be used to represent feature flag impressions.
+- **signal type for feature flag changes** - agree on which signal type or types should be used to represent feature flag changes.
+- **semantic conventions for feature flag impressions** - make any necessary additions to the semantic conventions to support feature flag impressions
+- **semantic conventions for feature flag changes** - make any necessary additions to the semantic conventions to support feature flag change events
+- **prototype feature flag impressions**
+- **prototype feature flag change events** - feature flag change events may not be generated by SDKs or instrumentations, but by the feature flag management tools.
```

Comment:
Just a minor comment on this particular point here.

If the management tooling raises the event, this could create a temporal inconsistency where some or all of the applications have not yet fetched the updated values.

In the .NET ecosystem (I'm speaking only on the platform I have experience with here, but this might apply to others), it is incredibly common for one to set a new value to a key (for example, on Azure AppConfiguration), but then have many applications waiting on different refresh timings or cache expirations for that key, meaning they would potentially still serve N requests using the previous value. Even when using something like a per-request middleware, the refresh operation tends to happen in the background and will usually not affect the _current_ request or other parallel requests happening around the same time.

If the only source for a flag change events is the central tool, one could end up checking behavior still running on the old value while the metric/etc indicates a new value should be present.

I think this needs to be reconsidered from the perspective of each application/process as well. Maybe we need a change event from the configuration source, and a second change event on each application when the new value is fetched? That to me would be the most comprehensive approach, although I'm not sure exactly how those event streams would be represented.

I almost want to suggest that OpenTelemetry should define a 4th type of telemetry called "events" (or "state transitions" or something), generalizing the concept and then making span events a particular use case of that. Then, when looking at a particular trace, one could include "global events" like these to understand the complete behavior on a given flow.

Anyways... just throwing some ideas around here.

Reply:
We planned to define semantics for change events from the management tool (source of truth) and the services (SDKs). That way, you could track how long it took for a change to propagate and quickly identify which services haven't been updated.

OpenTelemetry already has an [events concept](https://opentelemetry.io/docs/specs/otel/logs/event-api/) that's basically semantics on top of logs.

Reply:
> We planned to define semantics for change events from the management tool (source of truth) and the services (SDKs). That way, you could track how long it took for a change to propagate and quickly identify which services haven't been updated.

Perfect. Love to hear it.

> OpenTelemetry already has an [events concept](https://opentelemetry.io/docs/specs/otel/logs/event-api/) that's basically semantics on top of logs.

I remember seeing this but always assumed it was just a "synonym" to logs. Looks like I was wrong as it even has its own classes and abstractions.

Am I correct in that observability tools don't quite have this separation between logs and events yet though? Should they? We use Datadog extensively and I never saw this distinction anywhere over there (although I'm not necessarily an expert in DD either... so it could be me just missing it). 

I also don't know what exactly maps to these events from the .NET APIs, since we only have `ILogger` (for logs), `Meter` and `Instrument`s (for metrics) and `Activity` (for spans) there. The only place where I ever saw the concept of "events" was for "activity events", and those only make sense if you _are_ under an activity in the first place.

Reply:
>I remember seeing this but always assumed it was just a "synonym" to logs. Looks like I was wrong as it even has its own classes and abstractions.

Events and logs are the same data type. Events are LogRecords which have a name which uniquely defines a particular class or type of event. All events with the same name have Payloads that conform to the same schema, which assists in analysis in observability platforms. Events are described in more detail in the [semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/events/).

>Am I correct in that observability tools don't quite have this separation between logs and events yet though? Should they?

For the most part they don't. There could be arguments made either way as to if they should.

Reply:
Hello! I'm part of the `Microsoft.FeatureManagement` team.

> If the only source for a flag change events is the central tool, one could end up checking behavior still running on the old value while the metric/etc indicates a new value should be present.

We solve this with versioning emitted on the `feature_flag` event. For us- the events usually come from the SDK, but the centralized server dictates what version we're working with. This allows a given user/account/activity to be connected to the appropriate version of the flag, meaning you know which version applies to the rest of the metrics.

Reply:
Sadly, I'm unable to join at this time. Very interested to see what y'all land on.